### PR TITLE
Abstract EavtSink into TextSink and implement EavtTextSink

### PR DESCRIPTION
**Pull request by ewanv**
Allows definition of custom sinks using `Thrift` structs other than `Eavt`.

### CODE REVIEW

File: USERGUIDE.markdown [hunk 1]
```diff
@@ -286,6 +286,66 @@ object MultiPartitionSnippet {
...
+  implicit object FeatureEavtEnc extends FeatureValueEnc[FeatureEavt] {
+    def encode(fvt: (FeatureValue[_], Time)): FeatureEavt = fvt match {
+      case (fv, time) =>
+        val featureValue = (fv.value match {
+          case Integral(v) => v.map(_.toString)
+          case Decimal(v) => v.map(_.toString)
+          case Str(v) => v
+        }).getOrElse(TextSink.NullValue)
+
+        // TODO: Does time format need to be configurable?
```
Comment (dkristian) [inline USERGUIDE.markdown]: This comment can be removed

File: USERGUIDE.markdown [hunk 2]
```diff
@@ -177,7 +177,7 @@ Coppersmith currently supports the following source types:
 
 and one sink type:
 
-- `EavtSink`: a file in the format required for ingestion into the feature store (
+- `EavtTextSink`: a file in the format required for ingestion into the feature store (
```
Comment (dkristian) [inline USERGUIDE.markdown]: While we're here, we can ditch the reference to feature store:
`EavtTextSink`: writes EAVT records to a Hive-partitioned directory structure using delimited text encoding

File: USERGUIDE.markdown [hunk 3]
```diff
@@ -286,6 +286,66 @@ object MultiPartitionSnippet {
 }
 ```
 
+### Alternate sinks
+
+If an output format different to `EavtTextSink` is required, then `TextSink`
+should be used. A `Thrift` defining the sink format is needed, as well as
```
Comment (dkristian) [inline USERGUIDE.markdown]: A `Thrift` *struct* defining...

File: scalding/src/main/scala/commbank/coppersmith/scalding/EavtTextSink.scala [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import org.apache.hadoop.fs.Path
+import org.joda.time.DateTime
+
+import au.com.cba.omnia.maestro.api._, Maestro._
+
+import commbank.coppersmith.{FeatureValue, Feature}, Feature._, Value._
+import commbank.coppersmith.thrift.Eavt
+
+/**
+  * Created by vaughaew on 4/05/2016.
```
Comment (dkristian) [inline scalding/src/main/scala/commbank/coppersmith/scalding/EavtTextSink.scala]: Remove (inferable from git history)

File: scalding/src/main/scala/commbank/coppersmith/scalding/EavtTextSink.scala [hunk 5]
```diff
@@ -0,0 +1,49 @@
+package commbank.coppersmith.scalding
```
Comment (dkristian) [inline scalding/src/main/scala/commbank/coppersmith/scalding/EavtTextSink.scala]: Missing copyright

File: scalding/src/main/scala/commbank/coppersmith/scalding/FeatureSink.scala [hunk 6]
```diff
@@ -103,26 +101,29 @@ final case class DerivedSinkPartition[T, PP : PathComponents : TupleSetter](
   def tupleSetter = implicitly
 }
 
-object EavtSink {
+trait Encode[D, E] {
```
Comment (dkristian) [inline scalding/src/main/scala/commbank/coppersmith/scalding/FeatureSink.scala]: I realise I was the one that sent you this code, but this type is a bit general to have here. It looks like we could use `com.twitter.util.Encoder` to get similar (and more) functionality.

Comment (dkristian) [inline scalding/src/main/scala/commbank/coppersmith/scalding/FeatureSink.scala]: I realise I was the one that sent you this code, but this type is a bit general to have here. It looks like we could use `com.twitter.util.Encoder` to get similar (and more) functionality.

File: scalding/src/test/scala/commbank/coppersmith/scalding/TestEavtTextSink.scala [hunk 7]
```diff
@@ -0,0 +1,32 @@
+package commbank.coppersmith.scalding
```
Comment (dkristian) [inline scalding/src/test/scala/commbank/coppersmith/scalding/TestEavtTextSink.scala]: Missing copyright notice, and superfluous "Created" scaladoc comment below.

File: scalding/src/test/scala/commbank/coppersmith/scalding/TextSinkSpec.scala [hunk 8]
```diff
@@ -30,11 +30,12 @@ import au.com.cba.omnia.thermometer.core.Thermometer._
 import au.com.cba.omnia.thermometer.fact.PathFactoids.{exists, missing, records}
 import au.com.cba.omnia.thermometer.hive.ThermometerHiveSpec
 
-import commbank.coppersmith._, Arbitraries._, Feature.Value
+import commbank.coppersmith._, Arbitraries._, commbank.coppersmith.Feature.Value
```
Comment (dkristian) [inline scalding/src/test/scala/commbank/coppersmith/scalding/TextSinkSpec.scala]: Unnecessary change?

File: CHANGELOG.markdown [hunk 9]
```diff
@@ -1,6 +1,17 @@
 Change log
 ==========
 
+## 0.11.0
+Allow custom sinks.
```
Comment (dkristian) [inline CHANGELOG.markdown]: This change is more about supporting arbitrary feature value serialisation (custom sinks were possible before, but required more boiler plate).

**Comment by dkristian** [comment]
Good stuff. :+1: once comments are addressed

**Comment by ewanv** [comment]
Fixes pushed. I've changed a bit more of the CHANGELOG update, the previous type passed to `encode` was incorrect.

**Comment by dkristian** [comment]
:+1: good to go
